src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/JsonCollectionConverter.cs

```diff
@@ -12,41 +13,12 @@ namespace System.Text.Json.Serialization
     /// Base class for all collections. Collections are assumed to implement <see cref="IEnumerable{T}"/>
     /// or a variant thereof e.g. <see cref="IAsyncEnumerable{T}"/>.
     /// </summary>
-    internal abstract class JsonCollectionConverter<TCollection, TElement> : JsonResumableConverter<TCollection>
+    internal abstract class JsonCollectionConverter<TCollection, TElement, IntermediateType> : JsonAdvancedConverter<TCollection, IntermediateType>
     {
-        internal override bool SupportsCreateObjectDelegate => true;
         private protected sealed override ConverterStrategy GetDefaultConverterStrategy() => ConverterStrategy.Enumerable;
         internal override Type ElementType => typeof(TElement);
 
-        protected abstract void Add(in TElement value, ref ReadStack state);
-
-        /// <summary>
-        /// When overridden, create the collection. It may be a temporary collection or the final collection.
-        /// </summary>
-        protected virtual void CreateCollection(ref Utf8JsonReader reader, scoped ref ReadStack state, JsonSerializerOptions options)
-        {
-            JsonTypeInfo typeInfo = state.Current.JsonTypeInfo;
-
-            if (typeInfo.CreateObject is null)
-            {
-                // The contract model was not able to produce a default constructor for two possible reasons:
-                // 1. Either the declared collection type is abstract and cannot be instantiated.
-                // 2. The collection type does not specify a default constructor.
-                if (TypeToConvert.IsAbstract || TypeToConvert.IsInterface)
-                {
-                    ThrowHelper.ThrowNotSupportedException_CannotPopulateCollection(TypeToConvert, ref reader, ref state);
-                }
-                else
-                {
-                    ThrowHelper.ThrowNotSupportedException_DeserializeNoConstructor(TypeToConvert, ref reader, ref state);
-                }
-            }
-
-            state.Current.ReturnValue = typeInfo.CreateObject();
-            Debug.Assert(state.Current.ReturnValue is TCollection);
-        }
-
-        protected virtual void ConvertCollection(ref ReadStack state, JsonSerializerOptions options) { }
+        private protected abstract void Add(ref IntermediateType collection, in TElement value, JsonTypeInfo collectionTypeInfo);
```

Comment:
I think once we get to the point of bringing collection support to contract customization, I would imagine we would be exposing a `Action<object, object>? Add { get; set; }` to JsonTypeInfo. Perhaps we could abstract away the concept of intermediate accumulators by having all the collection lambdas accept an additional `object? state` parameter.

Reply:
By pushing all these concerns to `JsonTypeInfo`, it might be that we could replace all enumerable converter types with just one that calls into delegates. I expect this might offer even more size reduction benefits.